lib/gc/Analysis/MatmulConfigAnalysis.cpp

```diff
@@ -0,0 +1,422 @@
+  uint32_t MTaskNum = llvm::divideCeil(M, config.MBlock);
+  uint32_t NTaskNum = llvm::divideCeil(N, config.NBlock);
+  uint32_t KTaskNum = llvm::divideCeil(K, config.KBlock);
+  double cost = (MTaskNum % config.MThreads) * 1.0 / MTaskNum +
+                (NTaskNum % config.NThreads) * 1.0 / NTaskNum +
+                (KTaskNum % config.KThreads) * 1.0 / KTaskNum;
+  if (MTaskNum < config.MThreads || NTaskNum < config.NThreads ||
+      KTaskNum < config.KThreads) {
+    double threadNotFulllyUtilizedPenalty = 10.0;
+    cost *= threadNotFulllyUtilizedPenalty;
+  }
+  return cost;
+}
+
+constexpr unsigned bitPerByte = 8;
+// calculate the cost of the memory consumption on the thread
+double memoryConsumptionOnThreadCost(linalg::LinalgOp &linalgOp,
+                                     ArrayRef<uint32_t> shape,
+                                     const MatmulConfig &config,
+                                     SystemDesc &sysDesc) {
+  uint32_t M = shape[0], N = shape[1], K = shape[2];
+  llvm::TypeSize dtypeSize = DataLayout().getTypeSizeInBits(
+      ShapeAdaptor(linalgOp.getDpsInputs()[1].getType()).getElementType());
+  // if use K split, there will be one more final reduce and break the post
+  // fusion
+  double KSplitPenalty = 8.0 * (dtypeSize / bitPerByte);
+  double memoryConsumptionPerThread =
+      M * K * 1.0 / config.MThreads / config.KThreads +
+      K * N * 1.0 / config.KThreads / config.NThreads +
+      M * N * ((config.KThreads - 1) * KSplitPenalty + 1.0) / config.MThreads /
+          config.NThreads;
+  return memoryConsumptionPerThread;
+}
+
+// calculate the cost of the computation intensity on the L2 cache
+double computationIntensityOnL2Cache(linalg::LinalgOp &linalgOp,
+                                     ArrayRef<uint32_t> shape,
+                                     const MatmulConfig &config,
+                                     SystemDesc &sysDesc) {
+  double fullLoadRatio = 0.7;
+  uint32_t L2Cache = sysDesc.getCacheSize(2);
+  llvm::TypeSize dtypeSize = DataLayout().getTypeSizeInBits(
+      ShapeAdaptor(linalgOp.getDpsInputs()[1].getType()).getElementType());
+  uint32_t outOfCachePenalty = 1024;
+  double FLOPS = 2.0 * config.MBlock * config.NBlock * config.KBlock;
+  double memoryConsumption = config.MBlock * config.NBlock +
+                             config.NBlock * config.KBlock +
+                             config.MBlock * config.KBlock;
+  double computationIntensity = FLOPS / memoryConsumption;
+  if (memoryConsumption * (dtypeSize / bitPerByte) > L2Cache * fullLoadRatio) {
+    computationIntensity /= outOfCachePenalty;
+  }
+  return 1 / computationIntensity;
+}
+
+using CostModelFn =
+    std::function<float(linalg::LinalgOp &linalgOp, ArrayRef<uint32_t> shape,
+                        MatmulConfig cfg, SystemDesc &sysDesc)>;
+
+// filter the config by the cost model
+std::vector<MatmulConfig>
+filterConfigByCostModel(std::vector<MatmulConfig> configs,
```

Comment:
Using const reference to avoid extra copy or mutation?

Reply:
Thanks! Replace it with arrayref